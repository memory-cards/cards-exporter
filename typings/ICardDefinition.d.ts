export interface ICardDefinition {
  type: string;
  lang: string;
  tags: string[];
  card: {
    question: string;
    comment?: string;
    answers?: [string];
  };
}
