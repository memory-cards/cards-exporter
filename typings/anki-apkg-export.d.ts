// @todo : add declaration

declare module "anki-apkg-exportoo" {
  const x: any;
  export = x;
}
