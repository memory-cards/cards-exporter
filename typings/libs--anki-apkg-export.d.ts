declare class AnkiExport {
  public constructor(name: string);

  public addCard(front: string, back: string, tags?: string[]);

  public save(): Promise<string | Buffer>;
}

declare module "anki-apkg-export" {
  export = AnkiExport;
}
