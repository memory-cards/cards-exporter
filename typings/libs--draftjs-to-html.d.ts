function draftToHtml(draft: any): string;

declare module "draftjs-to-html" {
  export = draftToHtml;
}
