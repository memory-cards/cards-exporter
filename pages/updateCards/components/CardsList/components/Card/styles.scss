.card {
  background-color: #ffffff;
  margin: 0 0 5px;
  padding: 5px;
  height: 40px;
}
