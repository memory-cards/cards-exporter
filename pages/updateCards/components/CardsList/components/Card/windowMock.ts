window.document.createElement = jest.fn().mockImplementation(() => {
  const element = {};

  Object.defineProperty(element, "innerHTML", {
    /* tslint:disable object-literal-shorthand */
    set: function(x) {
      this.innerText = x;
    }
  });

  return element;
});
