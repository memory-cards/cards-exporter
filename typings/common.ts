export interface RepoTags {
  [tagName: string]: number;
}

export enum CardType {
  INFO = "info",
  CHOOSE_SEQUENCE = "choose_sequence",
  CHOOSE_OPTIONS = "choose_options",
  ORDER_ITEMS = "order_items"
}
