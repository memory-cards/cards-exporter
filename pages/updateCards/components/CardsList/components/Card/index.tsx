import React from "react";
import { ICardDefinition } from "~/typings/ICardDefinition";

import "./styles.scss";

interface Props {
  card: ICardDefinition;
  selectCard: (selectedCard: ICardDefinition) => void;
}

const Card = ({ card, selectCard }: Props) => {
  const onClick = () => {
    selectCard(card);
  };

  const getShortCardQuestion = () => {
    const element = document.createElement("div");
    element.innerHTML = card.card.question;
    const text = element.innerText
      .split(" ")
      .slice(0, 10)
      .map(word => (word.length < 30 ? word : word.substr(0, 30)))
      .join(" ");
    return text.length < element.innerText.length ? `${text}...` : text;
  };

  return (
    <li className="card" onClick={onClick}>
      <div>{getShortCardQuestion()}</div>
      <br />
      <div>
        {card.tags.map(tag => (
          <span className="tag">{tag}</span>
        ))}
      </div>
    </li>
  );
};

export default Card;
