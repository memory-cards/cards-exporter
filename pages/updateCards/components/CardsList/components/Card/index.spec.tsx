/**
 * @jest-environment jsdom
 */
import { shallow, ShallowWrapper } from "enzyme";
import * as React from "react";
import * as TestRenderer from "react-test-renderer";
import { ICardDefinition } from "~/typings/ICardDefinition";

import "../Card/windowMock";

import Card from "./index";

const cardQuestion = Array(20)
  .fill("1")
  .join(" ");

const card = {
  card: {
    comment: "comment",
    question: cardQuestion
  },
  lang: "en",
  tags: ["js", "ts"],
  type: "type1"
};

describe("With Enzyme", () => {
  let component: ShallowWrapper;
  let selectCard: (selectedCard: ICardDefinition) => void;

  beforeEach(() => {
    selectCard = jest.fn();
    component = shallow(
      <Card card={card} selectCard={selectCard} isSelected={true} />
    );
  });

  describe("Card", () => {
    it("has correct classname for selected card", () => {
      expect(component.find(".selected").exists()).toBe(true);
    });

    it("has correct classname for not selected card", () => {
      component.setProps({ isSelected: false });

      expect(component.find(".selected").exists()).toBe(false);
    });

    it("renders tags labels", () => {
      const tags = component.find("span.badge");
      expect(tags).toHaveLength(card.tags.length);
      tags.forEach((tag, index) => expect(tag.text()).toBe(card.tags[index]));
    });

    it("renders only first 10 words of the question as card label", () => {
      const shortQuestion = component
        .find("div")
        .at(0)
        .text();
      const expectedText = `${Array(10)
        .fill("1")
        .join(" ")}...`;

      expect(shortQuestion).toBe(expectedText);
    });

    it("renders only shortened words in card label", () => {
      const newCardQuestion = Array(20)
        .fill("5".repeat(40))
        .join(" ");

      const newCard = {
        ...card,
        card: { ...card.card, question: newCardQuestion }
      };

      component = shallow(
        <Card card={newCard} selectCard={selectCard} isSelected={false} />
      );

      const cardQuestionLabel = component
        .find("div")
        .at(0)
        .text();

      const expectedText = `${Array(10)
        .fill("5".repeat(30))
        .join(" ")}...`;

      expect(cardQuestionLabel).toBe(expectedText);
    });

    it("renders short card question without dots if it's not big", () => {
      const newCardQuestion = Array(5)
        .fill("5")
        .join(" ");

      const newCard = {
        ...card,
        card: { ...card.card, question: newCardQuestion }
      };

      component = shallow(
        <Card card={newCard} selectCard={selectCard} isSelected={false} />
      );

      const cardQuestionLabel = component
        .find("div")
        .at(0)
        .text();

      expect(cardQuestionLabel).toBe(newCardQuestion);
    });

    it("click on card calls selectCard method", () => {
      component.find("li").simulate("click");

      expect(selectCard).toHaveBeenCalled();
      expect(selectCard).toHaveBeenCalledWith(card);
    });
  });
});

describe("With Snapshot Testing", () => {
  it('IndexPage shows "Cards Exporter"', () => {
    const component = TestRenderer.create(
      <Card card={card} selectCard={jest.fn()} isSelected={false} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
