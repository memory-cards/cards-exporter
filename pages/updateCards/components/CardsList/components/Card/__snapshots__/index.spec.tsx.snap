// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With Snapshot Testing IndexPage shows "Cards Exporter" 1`] = `
<li
  className="card-item"
  onClick={[Function]}
>
  <div>
    1 1 1 1 1 1 1 1 1 1...
  </div>
  <br />
  <div>
    <span
      className="tag"
    >
      js
    </span>
    <span
      className="tag"
    >
      ts
    </span>
  </div>
</li>
`;
